<a href="https://colab.research.google.com/github/paulIruaRosero/Paul-Irua-Rosero-Se-ales-y-Sistemas/blob/main/Ejercicios/3_loops.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Bucles

En Python, tenemos dos estructuras para bucles: el bucle for y el bucle while (no existe un bucle do- while).

Es importante tener claro los conceptos de contador y acumulador, ya que son fundamentales a la hora de trabajar con cualquier tipo de bucle.

Un contador es una variable que realiza un seguimiento del número de veces que ocurre un evento o condición en particular. Se usa comúnmente en bucles para controlar el proceso de iteración. El contador generalmente se inicializa antes de que comience el ciclo y se incrementa o disminuye con cada iteración.

Un acumulador, por otro lado, es una variable que acumula o recopila valores a lo largo de múltiples iteraciones de un bucle. A menudo se utiliza para calcular una suma, un producto o cualquier otro valor agregado en función de los valores obtenidos durante cada iteración del ciclo. El acumulador normalmente se inicializa antes del bucle y se actualiza dentro del cuerpo del bucle.

Tanto los contadores como los acumuladores desempeñan funciones cruciales en el control y la gestión del flujo de un bucle, lo que le permite realizar acciones específicas basadas en el estado actual o el progreso de la iteración del bucle.

# While

También se llaman bucles indeterminados, y se utilizan cuando sabemos la condición que se debe cumplir para finalizar el bucle, pero no sabemos cuántas veces ni cuánto tiempo va a tardar.

In [ ]:
condition = 1
while condition:
  print('Instructions inside the loop')
  condition = 0
print('Instructions outside the loop')

Instructions inside the loop
Instructions outside the loop


La estructura del bucle es muy sencilla, ya que lo único que hay que especificar es la condición que se debe cumplir para continuar en el bucle. Esencialmente, funciona de manera similar a una declaración if() condicional, excepto que el código dentro del bucle se ejecutará tantas veces como sea necesario siempre que la condición sea verdadera.

Dado que el bucle puede entenderse como una declaración condicional, funciona de la misma manera para los valores booleanos, así como para los valores 1 y 0 explicados en otros materiales.

In [ ]:
a = 1
while a:
  print("Choose an option:\n1. ---\n2. ---\n0. Exit\n\n")
  a = int(input())


Choose an option:
1. ---
2. ---
0. Exit


1
Choose an option:
1. ---
2. ---
0. Exit


2
Choose an option:
1. ---
2. ---
0. Exit


0


En el ejemplo anterior, nos basamos en el valor de 'a' como condición para el bucle.

Sin embargo, asignar un valor de antemano a veces puede causar problemas.

Otra forma de hacerlo es realizando directamente una comparación como condición del bucle.

In [ ]:
number = int(input('Enter an integer number. 0 to finish: '))

while number != 0:
    number = int(input('Enter an integer number. 0 to finish: '))

print('End of the program.')

Enter an integer number. 0 to finish: 5
Enter an integer number. 0 to finish: 0
End of the program.


Entonces, es posible utilizar bucles infinitos como while(True) o while(1) si es necesario que algo se ejecute todo el tiempo.

Aunque no es ideal usarlo con el bucle while(), existe la sentencia break, que permite salir del bucle en los escenarios anteriores, en caso de que sea crítico no ejecutar las líneas restantes antes de finalizar el bucle. .

Un error común al usar bucles while() consiste en no actualizar la variable de control, creando así un bucle infinito no intencional (un error en el código).

In [ ]:
i = 0

# Run this loop as long as i is less than 15
while i < 15:
    # Show a message
    print("¡Hello world!")
    i+=1



¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!


In [ ]:
4%2 #the % operator give you the residual between the numbers!

0

# For

Los bucles for() se utilizan para tener un control total de cuántas iteraciones se realizarán.

Es necesario tener un valor "inicial" y un valor "final".

Opcionalmente podemos tener un tamaño de paso determinado.

A diferencia de C, por ejemplo, donde tenemos todos estos argumentos en la declaración de bucle, en Python tenemos que usar la función [rango](https://www.w3schools.com/python/ref_func_range.asp).

In [ ]:
#for counter_variable in range(initial_value, final_value, step_size):

    # code within the loop


# code out of the loop

Por ejemplo, si queremos que un código muestre todos los números pares del 900 al 1000, tenemos:

In [ ]:
for i in range(900, 1000, 2):
    print(i)

900
902
904
906
908
910
912
914
916
918
920
922
924
926
928
930
932
934
936
938
940
942
944
946
948
950
952
954
956
958
960
962
964
966
968
970
972
974
976
978
980
982
984
986
988
990
992
994
996
998


Podemos ver como imprimió hasta 998 (como se mostró, tomamos hasta un número antes; si queremos incluir 1000, debe ser al menos hasta 1001).


Podemos aprovechar los argumentos predeterminados de la función range() para el siguiente ejemplo, donde tomamos los números del 0 al 10000 y contamos todos los que son múltiplos de 33.

In [ ]:
counter = 0

for i in range(10000):
    if (i % 33 == 0): # residual == 0?
        counter += 1

#show counter
print(counter)

304


Los bucles For se utilizan por excelencia para recorrer arrays, listas o matrices, ya que sus dimensiones son valores fijos y permiten iterar sus elementos a voluntad.

In [ ]:
a = [8, 10, 23, 17, 15] #list of numbers

for i in range(len(a)):
  print(a[i])

8
10
23
17
15


Es importante notar cómo len(a) devuelve el número de elementos, pero en Python, el índice va desde cero, es por eso que se imprimieron todos los elementos. Veamos la diferencia:

In [ ]:
len(a)

5

In [ ]:
for i in range(4):
  print(a[i])

8
10
23
17


In [ ]:
for i in range(5):
  print(a[i])

8
10
23
17
15


Luego, como tenemos condicionales anidados, también podemos tener bucles anidados, donde cada iteración de los bucles externos constituye una iteración completa del bucle interno. Combinándolo con lo dicho antes, veamos un ejemplo con una matriz (lista de listas):

In [ ]:
M = [[1,3,4],[5,6,7],[8,9,2]]
M

[[1, 3, 4], [5, 6, 7], [8, 9, 2]]

In [ ]:
for i in range(3): # Rows
  for j in range(3): # Columns
    print("Row " + str(i) + ", Column " + str(j) + ": " + str(M[i][j]))

Row 0, Column 0: 1
Row 0, Column 1: 3
Row 0, Column 2: 4
Row 1, Column 0: 5
Row 1, Column 1: 6
Row 1, Column 2: 7
Row 2, Column 0: 8
Row 2, Column 1: 9
Row 2, Column 2: 2


# Ejemplos

Escriba un programa que imprima los números del 1 al 100, pero para múltiplos de tres, imprima "Fizz" en lugar del número, y para múltiplos de cinco, imprima "Buzz".

Para números que son múltiplos de tres y cinco, imprima "FizzBuzz"

In [ ]:
for i in range(1, 101):
    if i % 3 == 0 and i % 5 == 0: #and operator included
        print("FizzBuzz")
    elif i % 3 == 0:
        print("Fizz")
    elif i % 5 == 0:
        print("Buzz")
    else:
        print(i)

1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz
16
17
Fizz
19
Buzz
Fizz
22
23
Fizz
Buzz
26
Fizz
28
29
FizzBuzz
31
32
Fizz
34
Buzz
Fizz
37
38
Fizz
Buzz
41
Fizz
43
44
FizzBuzz
46
47
Fizz
49
Buzz
Fizz
52
53
Fizz
Buzz
56
Fizz
58
59
FizzBuzz
61
62
Fizz
64
Buzz
Fizz
67
68
Fizz
Buzz
71
Fizz
73
74
FizzBuzz
76
77
Fizz
79
Buzz
Fizz
82
83
Fizz
Buzz
86
Fizz
88
89
FizzBuzz
91
92
Fizz
94
Buzz
Fizz
97
98
Fizz
Buzz


Escriba un programa que solicite al usuario que ingrese un número entero positivo y luego calcule la suma de todos los números enteros desde 1 hasta el número ingresado.

In [ ]:
num = int(input("Enter a positive integer: "))
sum = 0

# check if num is positive
if num <= 0:
    print("Invalid input! Please enter a positive integer.")
else:
    # calculate the sum of integers from 1 to num
    for i in range(1, num+1):
        sum += i

    print("The sum of integers from 1 to {} is: {}".format(num, sum))

Enter a positive integer: 6
The sum of integers from 1 to 6 is: 21


Genera un número aleatorio entre 1 y 100 y solicita al usuario que adivine el número.

Si la suposición es demasiado alta, el programa debería imprimir "Demasiado alto", si la suposición es demasiado baja, el programa debería imprimir "Demasiado bajo" y si la suposición es correcta, el programa debería imprimir "¡Felicitaciones!" y salir del bucle:

In [ ]:
import random

# generate a random number between 1 and 100
num = random.randint(1, 100)

# loop until the user guesses the number
while True:
    guess = int(input("Guess the number (between 1 and 100): "))

    # check if the guess is correct
    if guess == num:
        print("Congratulations! You guessed the number.")
        break

    # check if the guess is too high
    elif guess > num:
        print("Too high! Guess again.")

    # check if the guess is too low
    else:
        print("Too low! Guess again.")

Guess the number (between 1 and 100): 50
Too low! Guess again.
Guess the number (between 1 and 100): 80
Too low! Guess again.
Guess the number (between 1 and 100): 90
Too low! Guess again.
Guess the number (between 1 and 100): 20
Too low! Guess again.
Guess the number (between 1 and 100): 1
Too low! Guess again.
Guess the number (between 1 and 100): 99
Congratulations! You guessed the number.


Escriba un programa que genere una tabla de multiplicar para todos los números entre 1 y 10. (consulte el método [.format()](https://www.geeksforgeeks.org/python-string-format-method/))

In [ ]:
for i in range(1, 11):
    for j in range(1, 11):
        print("{:4d}".format(i*j), end="")
    print()

   1   2   3   4   5   6   7   8   9  10
   2   4   6   8  10  12  14  16  18  20
   3   6   9  12  15  18  21  24  27  30
   4   8  12  16  20  24  28  32  36  40
   5  10  15  20  25  30  35  40  45  50
   6  12  18  24  30  36  42  48  54  60
   7  14  21  28  35  42  49  56  63  70
   8  16  24  32  40  48  56  64  72  80
   9  18  27  36  45  54  63  72  81  90
  10  20  30  40  50  60  70  80  90 100


## Ejecicio 1

Escriba un código que le solicite al usuario que ingrese un número entero cada vez, y el proceso se repetirá con más números hasta que el usuario ingrese un cero.

Al final (cuando se ingresa el cero), muestra la cantidad de números ingresados.

In [ ]:
total_numeros = [] #creamos una lista donde vamos a almacenar los enteros

while True:
    num_entero = int(input('Por favor ingrese un número entero (ingrese 0 para terminar): '))

    if num_entero == 0:
        break  # Si el número ingresado es 0, el programa termina

    total_numeros.append(num_entero)


# Al final, se muestra la lista completa
print('Fin del programa.')
print('Los números ingresados fueron:', total_numeros)


Por favor ingrese un número entero (ingrese 0 para terminar): 4
Por favor ingrese un número entero (ingrese 0 para terminar): 3
Por favor ingrese un número entero (ingrese 0 para terminar): 5
Por favor ingrese un número entero (ingrese 0 para terminar): 0
Fin del programa.
Los números ingresados fueron: [4, 3, 5]


## Ejercicio 2
Utilice la función input() para pedirle al usuario que ingrese un número entero. Sólo si el número es par se acumulará.

El programa debe finalizar si la suma de los números supera 50 o si el usuario introduce un máximo de 10 números.

In [ ]:
print('Bienvenido al programa.')
lista1 = []

while True:
    num1 = int(input('por favor ingrese un número entero(ingrese 0 para terminar): '))
    if num1 == 0:
        break
    if num1 % 2 == 0:
      lista1.append(num1)
    if sum(lista1) > 50 or len(lista1) >= 10:
        break

print('Fin del programa.')
print('Los números pares ingresados fueron: ', lista1)

Bienvenido al programa.
por favor ingrese un número entero(ingrese 0 para terminar): 4
por favor ingrese un número entero(ingrese 0 para terminar): 2
por favor ingrese un número entero(ingrese 0 para terminar): 6
por favor ingrese un número entero(ingrese 0 para terminar): 4
por favor ingrese un número entero(ingrese 0 para terminar): 2
por favor ingrese un número entero(ingrese 0 para terminar): 2
por favor ingrese un número entero(ingrese 0 para terminar): 2
por favor ingrese un número entero(ingrese 0 para terminar): 8
por favor ingrese un número entero(ingrese 0 para terminar): 6
por favor ingrese un número entero(ingrese 0 para terminar): 4
Fin del programa.
Los números pares ingresados fueron:  [4, 2, 6, 4, 2, 2, 2, 8, 6, 4]


2Developed by:

- Brian Patino Agudelo - bpatinoa@unal.edu.co

- Andrés Marino Álvarez Meza - amalvarezme@unal.edu.co

- César Germán Castellanos Domínguez - cgcastellanosd@unal.edu.co